cmd/pmm-agent-entrypoint/main.go

```diff
@@ -0,0 +1,271 @@
+The following help text shows them as [PMM_AGENT_XXX].
+`
+
+type RestartPolicy int
+
+const (
+	DoNotRestart RestartPolicy = iota + 1
+	RestartAlways
+	RestartOnFail
+)
+
+var (
+	pmmAgentSetupEnv        = getEnvWithDefault("PMM_AGENT_SETUP", "false")
+	pmmAgentSidecarEnv      = getEnvWithDefault("PMM_AGENT_SIDECAR", "false")
+	pmmAgentSidecarSleepEnv = getEnvWithDefault("PMM_AGENT_SIDECAR_SLEEP", "1")
+	pmmAgentPrerunFile      = getEnvWithDefault("PMM_AGENT_PRERUN_FILE", "")
+	pmmAgentPrerunScript    = getEnvWithDefault("PMM_AGENT_PRERUN_SCRIPT", "")
+)
+
+var (
+	pmmAgentProcessID int = 0
+)
+
+func getEnvWithDefault(key, defautlValue string) string {
+	if value, ok := os.LookupEnv(key); ok {
+		return value
+	}
+	return defautlValue
+}
+
+func init() {
+	logrus.SetFormatter(&logrus.TextFormatter{
+		ForceColors:     true,
+		FullTimestamp:   true,
+		TimestampFormat: "2006-01-02T15:04:05.000-07:00",
+	})
+}
+
+func runPmmAgent(commandLineArgs []string, restartPolicy RestartPolicy, l *logrus.Entry, pmmAgentSidecarSleep int) int {
+	pmmAgentFullCommand := "pmm-admin " + strings.Join(commandLineArgs, " ")
+	for {
+		l.Infof("Starting 'pmm-admin %s'...", strings.Join(commandLineArgs, " "))
+		cmd := commandPmmAgent(commandLineArgs)
+		if err := cmd.Start(); err != nil {
+			l.Errorf("Can't run: '%s', Error: %s", commandLineArgs, err)
+			return -1
+		}
+		var exitCode int
+		pmmAgentProcessID = cmd.Process.Pid
+		if err := cmd.Wait(); err != nil {
+			exitError, ok := err.(*exec.ExitError)
+			if !ok {
+				l.Errorf("Can't get exit code for '%d'. Error code: %s", pmmAgentFullCommand, err)
+				return -1
+			}
+			exitCode = exitError.ExitCode()
+		}
+		l.Infof("'%s' exited with %d", pmmAgentFullCommand, exitCode)
+
+		if restartPolicy == RestartAlways || (restartPolicy == RestartOnFail && exitCode != 0) {
+			l.Infof("Restarting `%s` in %d seconds because PMM_AGENT_SIDECAR is enabled...", pmmAgentFullCommand, pmmAgentSidecarSleep)
+			time.Sleep(time.Duration(pmmAgentSidecarSleep) * time.Second)
+		} else {
+			return exitCode
+		}
+
+	}
+}
+
+func commandPmmAgent(args []string) *exec.Cmd {
+	const pmmAgentCommandName = "pmm-agent"
+	command := exec.Command(pmmAgentCommandName, args...)
+	command.Stdout = os.Stdout
+	command.Stderr = os.Stderr
+	return command
+}
+
+func main() {
+	go reaper.Reap()
+
+	var status int
+
+	l := logrus.WithField("component", "entrypoint")
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
+
+	go func() {
+		s := <-signals
+		signal.Stop(signals)
+		l.Warnf("Got %s, shutting down...", unix.SignalName(s.(unix.Signal)))
+		if pmmAgentProcessID != 0 {
+			// gracefull shutdown for pmm-agent
+			if err := syscall.Kill(pmmAgentProcessID, syscall.SIGTERM); err != nil {
+				l.Warn("Failed to send SIGTERM, command must have exited:", err)
+			}
+		}
+		cancel()
```

Comment:
We can't use `CommandContext` because it sends SIGKILL to subprocess. I used it for some command but not for `pmm-agent`

Reply:
and here we exactly send SIGKILL command.
https://github.com/percona/pmm-agent/pull/380/files#diff-362c1b9d32b751c3986e334de58ca698848712a7fa4cf893424e9deac20d59a3R117-R123
So I suggested to use context instead of trying to extract PID and sending signal manually